WIP: Refactor setup and teardown methods in test classes

michizhou opened the pull request:
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #87 **:

**Description of the changes being introduced by the pull request**:
Made some refactoring changes of the setup/teardown methods of several test script files to simplify and centrally integrate their functions. Created a superclass (inherits from Unittest.TestCase) that overrides large portions of the setupclass/setup/teardownclass/teardown methods on it.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

SantiagoTorres reviewed (commented):
Hi, before I can review this I'd like you to fix the stray merge error markers.

michizhou commented:
How do I do that? I honestly have no idea on how to fix the merge error markers. I would appreciate it if you could give me a step-by-step guide on how to resolve them.

SantiagoTorres commented on tests/common.py:
```diff
@@ -29,6 +29,113 @@
+  sha256_foo_tar = \
+    "93c3c35a039a6a3d53e81c5dbee4ebb684de57b7c8be11b8739fd35804a0e918"
+
+  def setUpEnvironment(self, type=None, crypto=None):
```
I was hoping this was an override of `setUpClass` or the `setUp` method so that inner classes used this boilerplate automagically.

SantiagoTorres commented on tests/common.py:
```diff
@@ -29,6 +29,113 @@
+        os.path.dirname(os.path.realpath(__file__)), "gpg_keyrings", crypto)
+
+      shutil.copytree(keyring_path, self.gnupg_home)
+    elif type == 1:
```
I would like to use an actual type constant. It's not clear to me what `type` means in this context (I assume of a certain subgroup of the testCases with more commonalities than others.

JustinCappos replied:
Also type is a built in and you're variable shadowing here.  You want to pick a different name...

SantiagoTorres commented on tests/test_gpg.py:
```diff
@@ -63,24 +64,12 @@ class TestGPGRSA(unittest.TestCase):
-    shutil.copytree(gpg_keyring_path, self.gnupg_home)
-    os.chdir(self.test_dir)
-
+    common.TestCaseLib.setUpEnvironment("rsa")
```
Here I would've hoped for you to make it inherit from your new class (which in turn inherits from testCase)

SantiagoTorres commented on tests/test_gpg.py:
```diff
@@ -63,24 +64,12 @@ class TestGPGRSA(unittest.TestCase):
+    common.TestCaseLib.setUpEnvironment("rsa")
 
   @classmethod
   def tearDownClass(self):
```
same goes for here.

SantiagoTorres commented on tests/test_runlib.py:
```diff
@@ -504,35 +481,24 @@ def test_no_key_arguments(self):
   """"Test in_toto_record_stop(step_name, key, product_list). """
 
-  @classmethod
+  @classmethod    # test_runlib.py**
```
I'm not sure what this comment means.

SantiagoTorres reviewed (changes requested):
I see a couple of things that could improve. It's looking in the right direction though. Thanks!

SantiagoTorres commented:
@michizhou I see these builds are failing due to syntax errors. Do these pass on your machine?

michizhou commented:
@SantiagoTorres The thing is that the syntax errors are all related to how I am calling super() in the test script files to reference my superclass. On my end the syntax for my super() instances seems to be in line with the documentation for super().

SantiagoTorres commented:
The tests are still failing. Are you testing this locally before pushing?

michizhou commented:
The thing is I am not sure how to test locally, so I haven't been able to do it. Also, it seems that the checks are failing because of some variables being undefined (i.e. Layout, tempfile) as well as a small issue related to indent formatting and some others that I am not quite sure why are popping up (since I had fixed the issue of giving the setUpClass methods a 'self' argument). I was hoping that you could share your thoughts on why these errors may be happening as well as give me some tips on how to test these files on my local developer environment on my end.

SantiagoTorres commented:
There's a .DS_STORE on this PR too. You probably want to add these to your personal .gitignore.

SantiagoTorres commented on .travis.yml:
```diff
@@ -9,15 +9,12 @@ matrix:
         - python: "2.7"
           env: TOXENV=py27
+        - python: "3.4"
+          env: TOXENV=py34
```
I'm not entirely sure we need this env anymore @lukpueh any thoughts as to why this sneaked here?

michizhou replied:
This was just the result of a mis-merge, which I have already fixed. @lukpueh Don't worry about it, as I have already gotten rid of the Python 3.4 env in this file.

SantiagoTorres commented on CHANGELOG.md:
```diff
@@ -1,22 +1,5 @@
-* Improve gpg support
-* Add full support for ed25519 keys and add optional key type parameter
-* Fix bug in rule verification (https://github.com/in-toto/in-toto/pull/204)
-
```
I think there's a bunch of changes here that are a result of a mis-merge. I think you need to update your branch...

michizhou replied:
@SantiagoTorres Do you want me to revert to the old changes or add back in the changes made by @lukpueh with the info about in-toto v0.2.3? If it's the latter, I have already reverted CHANGELOG.md back to its state before his additions.

SantiagoTorres commented on appveyor.yml:
```diff
@@ -1,9 +1,5 @@
-    - PYTHON: "C:\\Python37"
-      PYTHON_VERSION: 3.7
-      PYTHON_ARCH: 32
-
```
Why did this happen?

michizhou replied:
@SantiagoTorres I had committed for another unrelated documentation issue but neglected to change branches before I did. I also wanted to sync my forked repo so as to keep it up-to-date with the original repo, hence the failed merge.

SantiagoTorres reviewed (changes requested):
I don't think I can finish reviewing with all of the noise from the other PR's. Can you rebase please?

lukpueh commented:
**Re testing locally:**

You have at least two options for testing locally. Either you run tox, which automatically creates a virtual environment for each Python version specified (see [`tox.ini:envlist`](https://github.com/in-toto/in-toto/blob/develop/tox.ini#L8)), installs in-toto and it's dependencies (see [`tox.ini:deps`](https://github.com/in-toto/in-toto/blob/develop/tox.ini#L11)), and runs the linters, unittests and codecoverage (see [`tox.ini:commands`](https://github.com/in-toto/in-toto/blob/develop/tox.ini#L17)):

```shell
# In project root (use virtualenv!)
pip install tox
tox
```

Or, in case you are working on a specific unit test and want to run it, you can do something like this (see [`unittest`](https://docs.python.org/3/library/unittest.html) for more info) :

```shell
# In project root (use virtualenv!)
# Install in-toto in development mode (plus ci deps)
pip install -e .
pip install -r requirements-ci.txt

# Run a single test
python tests/test_verifylib.py TestInTotoVerify.test_verify_passing
```

I hope this helps. @SantiagoTorres, maybe we should add something like this somewhere to our README?  Or maybe even better to the [`secure-systems-lab:dev-workflow`](https://github.com/secure-systems-lab/lab-guidelines/blob/master/dev-workflow.md)?  (ping @awwad). What do you think?

lukpueh commented:
**Re branch discipline**

We don't have a policy for [`git rebase` vs `git merge`](https://www.atlassian.com/git/articles/git-team-workflows-merge-or-rebase) but I tend to favor `rebase`, definitely for keeping my working branch clean (helpful for reviews), but also to resolve conflicts with the branch I want to pull into. @SantiagoTorres, @awwad, what's your stance on this?

@michizhou, one thing I noted is that your commit messages could definitely carry some more context. You can find more information in our [dev](https://github.com/secure-systems-lab/lab-guidelines/blob/master/dev-workflow.md) and  [commit](https://github.com/secure-systems-lab/lab-guidelines/blob/master/commits.md) guidelines.

awwad commented:
@lukpueh I am strongly in favor of rebase for keeping branches clean.  I advocate rebasing to avoid things like "Merge remote-tracking branch 'upstream/develop' into develop" in a PR that's being merged back upstream. It's confusing and makes the history harder to read.

lukpueh commented:
Thanks, @awwad! @michizhou, would you mind reading up a bit on `git rebase`? You could use this branch as playground to see if you can clean up the commit log and also update your commits while you're at it.

michizhou commented:
@lukpueh Sure thing! I will definitely look into git rebase and can play around with it a bit here.

michizhou commented:
@SantiagoTorres Are you available to review my latest changes to this PR? Now that the functionality of the refactored test scripts is assured, I would greatly appreciate it if I could get some constructive feedback so that I can further optimize and improve the flow of the test script code.

SantiagoTorres commented:
Hi!

yes I'll take a look at these soon. Cheers!

-Santiago.

On Fri, Oct 26, 2018 at 12:55:06PM -0700, michizhou wrote:
> @SantiagoTorres Are you available to review my latest changes to this PR? Now that the functionality of the refactored test scripts is assured, I would greatly appreciate it if I could get some constructive feedback so that I can further optimize and improve the flow of the test script code.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/in-toto/in-toto/pull/216#issuecomment-433525141

SantiagoTorres commented on tests/.idea/misc.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.6.2 (/Library/Frameworks/Python.framework/Versions/3.6/bin/python3.6)" project-jdk-type="Python SDK" />
+</project>
```
I think the `.idea` folder shouldn't live on the repository. Everyone uses their own editor.

SantiagoTorres commented on tests/bar:
```diff
@@ -0,0 +1 @@
+bar
```
This file doesn't seem to belong here.

SantiagoTorres commented on tests/common.py:
```diff
@@ -22,13 +22,110 @@
 from mock import patch
 
+class TestCaseLib(unittest.TestCase):
+  """TestCase subclass providing """
```
I think this needs some more text explaining what it is exactly providing.

SantiagoTorres commented on tests/common.py:
```diff
@@ -22,13 +22,110 @@
+  key = {}
+  key2 = None
+  key_path = "test_key"
+  key_path2 = "test-key2"
```
could we make these consistent? I'd err on having both filenames using underscores.

michizhou replied:
The original code had those inconsistent naming conventions, which I carried over. I'm worried that making both filenames use underscores may break the code, especially for the original `TestInTotoRecordStop` class in `test_runlib.py`.

lukpueh replied:
It's generally okay to also revise the tests and update them.

SantiagoTorres commented on tests/common.py:
```diff
@@ -22,13 +22,110 @@
+    self.test_dir = os.path.realpath(tempfile.mkdtemp())
+    os.chdir(self.test_dir)
+
+    if self.methodMode == 1:
```
could we use descriptive names for `methodMode` here please?

lukpueh replied:
... and snake_case instead of camelCase (see [style guideline](https://github.com/secure-systems-lab/code-style-guidelines#general-variable-names))

SantiagoTorres commented on tests/common.py:
```diff
@@ -22,13 +22,110 @@
+
+    if self.methodMode == 1:
+      generate_and_write_rsa_keypair(self.key_path)
+      # import pdb; pdb.set_trace()
```
this pdb statement crept in.

SantiagoTorres commented on tests/common.py:
```diff
@@ -22,13 +22,110 @@
+        generate_and_write_rsa_keypair(self.key_path2)
+        self.key2 = prompt_import_rsa_key_from_file(self.key_path2)
+
+      self.step_name = self.firstOpts
```
what does `firstOpts` mean? we probably want a descriptive name for this variable.

SantiagoTorres commented on tests/common.py:
```diff
@@ -22,13 +22,110 @@
+        self.key2 = prompt_import_rsa_key_from_file(self.key_path2)
+
+      self.step_name = self.firstOpts
+      if self.settingMode == 1:
```
Same for `settingMode`

SantiagoTorres commented on tests/common.py:
```diff
@@ -22,13 +22,110 @@
+    elif self.methodMode == 0:
+      if self.settingMode == 1:
+        # Create directory to run the tests without having everything blow up
+        self.gnupg_home = os.path.join(self.test_dir, self.firstOpts)
```
This doesn't quite create the directory though. Mind rephrasing the comment please?

michizhou replied:
What does this line do anyways? I'm not sure for certain what `gnupg_home` is for.

lukpueh replied:
`self.gnupg_home` points to the directory that has the gpg keychain. Try `grep` to see how `gnupg_home` is passed to in-toto functions, that should reveal its purpose.

SantiagoTorres commented on tests/common.py:
```diff
@@ -22,13 +22,110 @@
+    else:
+      os.chdir(self.working_dir)
+      shutil.rmtree(self.test_dir)
+      if extraMode:
```
what does `extraMode` mean?

SantiagoTorres commented on tests/foo:
```diff
@@ -0,0 +1 @@
+foo
```
this file doesn't belong. I'm worried that your tests may not be cleaning up properly?

SantiagoTorres commented on tests/test_param_substitution.py:
```diff
@@ -97,10 +99,11 @@ def test_substitute(self):
   def test_substitute_no_var(self):
     """Raise an error if the parameter is not filled-in"""
     with self.assertRaises(KeyError):
+      # import pdb; pdb.set_trace()
```
this crept in.

SantiagoTorres commented on tests/test_runlib.py:
```diff
@@ -99,20 +100,16 @@ def setUpClass(self):
-    self.working_dir = os.getcwd()
+    # mkdtemp uses $TMPDIR, which might contain a symlink
+    # but we want the absolute location instead
+    super(TestRecordArtifactsAsDict, self).setUpClass()
```
Is this comment related to this?

michizhou replied:
No it ain't. I already got rid of it.

SantiagoTorres commented on tests/test_runlib.py:
```diff
@@ -124,19 +121,9 @@ def setUpClass(self):
     """Clear the ARTIFACT_EXLCUDES after every test. """
-    in_toto.settings.ARTIFACT_EXCLUDE_PATTERNS = []
-    in_toto.settings.ARTIFACT_BASE_PATH = None
+    super(TestRecordArtifactsAsDict, self).tearDownClass(True)
```
I don't think you need this super call.

michizhou replied:
Should I just get rid of the entire `tearDown` method as well then?

SantiagoTorres commented on tests/test_settings.py:
```diff
@@ -17,7 +17,7 @@
 import unittest
 import in_toto.settings
-
+import common
```
is this import needed?

michizhou replied:
Not really. I have removed it accordingly.

SantiagoTorres commented on tests/test_verifylib.py:
```diff
@@ -85,30 +87,22 @@ def setUpClass(self):
     # find where the scripts directory is located.
     scripts_directory = os.path.join(
-        os.path.dirname(os.path.realpath(__file__)), "scripts")
+      os.path.dirname(os.path.realpath(__file__)), "scripts")
```
I think you can move this out of `setupClass` and remove `setupClass` altogether to avoid overridding it

lukpueh replied:
ping

SantiagoTorres commented on tests/test_verifylib.py:
```diff
@@ -299,19 +293,13 @@ class TestVerifyModifyRule(unittest.TestCase):
     self.materials = {
-      "foo": {"sha256": sha256_1},
-      "bar": {"sha256": sha256_1}
+      "foo": {"sha256": super(TestVerifyModifyRule, self).sha256_foo},
```
same here.

SantiagoTorres commented on tests/test_verifylib.py:
```diff
@@ -447,53 +435,41 @@ def test_fail(self):
+      "barfoo": {"sha256": super(TestVerifyMatchRule, self).sha256_barfoo},
+      "dev/bar": {"sha256": super(TestVerifyMatchRule, self).sha256_bar},
+      "dev/barfoo": {"sha256": super(TestVerifyMatchRule, self).sha256_barfoo},
+    }
```
There must be a more kosher way to initialize these dictionaries.

SantiagoTorres commented on tests/test_verifylib.py:
```diff
@@ -702,33 +678,28 @@ def test_fail_hash_not_equal(self):
+  links = {
+    "item": Metablock(signed=Link(name="item",
+        materials={
+            "foo": {"sha256": common.TestCaseLib.sha256_foo},
```
Interestingly, here you do it differently.

SantiagoTorres commented on tests/test_verifylib.py:
```diff
@@ -1457,36 +1386,26 @@ def test_verify_link_signature_thresholds__S_S_S(self):
+    super(self, TestVerifySublayouts).setUpClass()
 
     # copy portable scripts over
+    # import pdb; pdb.set_trace()
```
This crept in

SantiagoTorres commented on tests/test_verifylib.py:
```diff
@@ -1500,39 +1419,32 @@ def setUpClass(self):
     sub_layout = copy.deepcopy(layout_template)
     sub_layout_path = FILENAME_FORMAT.format(step_name=sub_layout_name,
-        keyid=alice_pub["keyid"])
+                                             keyid=alice_pub["keyid"])
```
This type of indentation is inconsistent with our coding guidelines.

SantiagoTorres reviewed (changes requested):
- [ ] I still see .DS_Store in here, could you remove them?
- [ ] There are a couple of test files and the `.idea` directory that shouldn't be in version control
- [ ] We may need to rename some variables so as to improve readability
- [ ] If possible, avoid using a custom setUp class with a call to `super` inside.
- [ ] Some code and constants could be moved to helped methods (e.g., the hashes that you are using super to get).

It's looking good though. Thanks!

michizhou commented:
@SantiagoTorres Okay so it looks like your suggestion about getting rid of the `setUpClass` method in `test_verifylib.py` by moving its contents out didn't work since it seems the particular order in which the lines are executed is important, with the super statement needed to navigate to the right directory in which the "foo" file is to be written after the script directory is found rather than before.

michizhou committed:
Refactored test script setup and teardown class methods

Changed implementation details for refactored setup class method processing that
is shared among test scripts. Also changed setup of refactored class variables
for specific test script files to ensure better readability.

lukpueh commented on tests/common.py:
```diff
@@ -22,13 +22,109 @@
   `python tests/runtests.py`.
 
 """
+
```
Please remove newline.

lukpueh commented on tests/common.py:
```diff
@@ -22,13 +22,109 @@
 
+class TestCaseLib(unittest.TestCase):
+  """TestCase subclass providing standard methods that will centrally
+  integrate test script setup and teardown methods through overriding"""
```
Do you think you can make this more precise? Think of someone, who is about to write a new test case and, after seeing that other test cases in the suite inherit from your `TestCaseLib` class, wants to find out what it can offer. :)

lukpueh commented on tests/test_param_substitution.py:
```diff
@@ -52,33 +54,33 @@
 
 
-class Test_SubstituteArtifacts(unittest.TestCase):
+class Test_SubstituteArtifacts(common.TestCaseLib):
```
Any reason why this is subclassing `TestCaseLib`? It isn't using any of its features.

lukpueh commented on tests/test_param_substitution.py:
```diff
@@ -52,33 +54,33 @@
+        "expected_products": [
+          ["CREATE", "{NEW_THING}"]
+        ]
+      }]
```
Why did you change the indentation here?

lukpueh replied:
Our style guideline suggests double indentation for line continuation

lukpueh commented on tests/test_param_substitution.py:
```diff
@@ -100,7 +102,7 @@ def test_substitute_no_var(self):
 
 
-class Test_SubstituteRunField(unittest.TestCase):
+class Test_SubstituteRunField(common.TestCaseLib):
```
See "why subclass TestCaseLib" comment above.

lukpueh commented on tests/test_param_substitution.py:
```diff
@@ -110,13 +112,13 @@ def setUp(self):
+        "name": "run-command",
+        "run": ["{COMMAND}"],
+      }]
+    })
```
See "why re-indent" comment above.

lukpueh commented on tests/test_param_substitution.py:
```diff
@@ -131,18 +133,18 @@ def test_inspection_fail_with_non_zero_retval(self):
 
 
-class Test_SubstituteExpectedCommand(unittest.TestCase):
+class Test_SubstituteExpectedCommand(common.TestCaseLib):
```
See "why subclass TestCaseLib" comment above.

lukpueh commented on tests/test_param_substitution.py:
```diff
@@ -131,18 +133,18 @@ def test_inspection_fail_with_non_zero_retval(self):
+      "steps": [{
+        "name": "run-command",
+        "expected_command": ["{EDITOR}"],
+      }]})
```
See "why re-indent" comment above.

lukpueh commented on tests/test_param_substitution.py:
```diff
@@ -157,45 +159,29 @@ def test_substitute_no_var(self):
+        "name": "run-command",
+        "expected_command": ["{EDITOR}"],
+      }]
+    })
```
See "why re-indent" comment above.

lukpueh commented on tests/test_param_substitution.py:
```diff
@@ -157,45 +159,29 @@ def test_substitute_no_var(self):
-  def tearDownClass(self):
-    os.chdir(self.working_dir)
-    shutil.rmtree(self.test_dir)
+      ["alice.pub"])
```
See "why re-indent" comment above.

lukpueh commented on tests/test_verifylib.py:
```diff
@@ -91,26 +93,18 @@ def setUpClass(self):
+        "name": "touch-bar",
+        "run": ["python", os.path.join(scripts_directory, "touch"), "bar"],
+      }]
+    })
```
See "why re-indent" comment above.

lukpueh commented on tests/common.py:
```diff
@@ -22,13 +22,109 @@
+        shutil.copytree(gpg_keyring_path, self.gnupg_home)
+      elif self.extra_settings == "demo":
+        # Copy demo files to temporary dir
+        for file in os.listdir(demo_files):
```
You're [name shadowing](https://github.com/in-toto/in-toto/pull/216#discussion_r217838750) again. Please never use reserved names (unless of course you need to override them). Any basic syntax highlighter should show you if you choose a reserved name, at least if it is a builtin.

lukpueh replied:
Also why not copying the entire directory using `shutil.copytree` and why not doing it above (in L90 below `if self.extra_settings == "demo":`)?

lukpueh commented on tests/common.py:
```diff
@@ -22,13 +22,109 @@
+      "d65165279105ca6773180500688df4bdc69a2c7b771752f0a46ef120b7fd8ec3"
+  sha256_2 = \
+      "cfdaaf1ab2e4661952a9dec5e8fa3c360c1b06b1a073e8493a7c46d2af8c504b"
+
```
Is there any reason to define these hashes as class variables here? `sha256_1` and `sha256_2` are only used in `TestVerifyItemRules` and the others are only used in `TestVerifyMatchRule`. Moreover, they don't correspond to any real file, they are only sample hashes used to test artifact rules. And if you take a closer look you'll see that `sha256_foo == sha256_1` and `sha256_bar == sha256_2`. So that's something to generalize. But that might as well happen somewhere else. E.g. as constants in `test_verifylib.py`.

lukpueh reviewed (changes requested):
Thanks for the changes, @michizhou. Generalizing common things from test cases, such as creating and changing to a temporary directory and/or copying demo files and/or copying gpg keychains to that temporary directory, is a very good idea!

However, I think your `TestCaseLib` combines a bit too many unrelated things, which has the odd side-effect that subclasses need to set variables like `need_key_pair`, `need_second` or `extra_settings`, in order to do case handling.

Here are a few suggestions:
 - If there are things that are only needed in one subclass you don't have to pull them up into the parent class.  You can still do sub-class specific work in the sub-class setup/teardown methods.
 - If things don't belong to the same class, just create different ones.  E.g. one that creates and changes to a tmp dir (in setUpClass) and changes back and removes the tmp dir (in tearDown). It could have child classes that e.g. copy over the demo files, or copy over the gpg keyring.
 - It might still be okay to provide some things in a parent class that are not necessarily needed in all child classes.

Two more general requests:
- Please don't commit trailing whitespace (we only have it as an [informal style guideline](https://github.com/secure-systems-lab/code-style-guidelines/issues/5), but I think I've asked you this before)
- Please don't use the `resolve conversation` button in GitHub. IMO it should be up to the reviewer/commenter to decide whether the raised concern has been resolved.

Let me know if you have any questions.

lukpueh commented:
Following above review and summarizing out-of-band discussions, here is what we plan for this PR: 

- Strip down `common.TestCaseLib` to serve one purpose only, i.e. creating and changing into a tempdir in setup and changing back to original `cwd` and deleting the tempdir in teardown. A more descriptive name in regards to that purpose would also make sense.
- Go through all `test_*.py` files and use that class where and as appropriate.

In follow-up PRs, we can add similar single-purpose classes in between that, e.g. copy demo files, or gpg keyrings, etc. ...

michizhou committed:
Changed content and scope of test script eafactoring methods

Simplified refactoring subclass to serve only one purpose: creating and changing
into temporary directory in setup and changing back to original current working
directory and deleting the temporary directory in teardown. Also included a more
descriptive name for subclass and modified its usage in test script files.

lukpueh commented on tests/common.py:
```diff
@@ -25,10 +25,42 @@
+  and flags which dictate execution of specialized code blocks.
+
+  Supports setup for key pair generation, demo supply chain link metadata,
+  key pair dummy material generation, and GPG keyring verification. Supports
```
Most of what you say in the docstring is inaccurate. There are no
*"flags which dictate execution of specialized code blocks"* and the class does not *"Supports setup for key pair generation, demo supply chain link metadata,  key pair dummy material generation, and GPG keyring verification"*. Please remove those parts.

lukpueh commented on tests/common.py:
```diff
@@ -25,10 +25,42 @@
 import unittest
 from mock import patch
 
+class SetupTestCase(unittest.TestCase):
```
As discussed, there may be multiple `TestCase`-setup superclasses (aka. Mixins). I'd say we make the class name more specific, e.g. `TempDirTestCase`

lukpueh commented on tests/test_verifylib.py:
```diff
@@ -1182,7 +1157,7 @@ def _verify_link_signature_tresholds(self, sig_id, auth_id, key_id):
     """Normal scenario. """
     layout, chain_link_dict = self._verify_link_signature_tresholds(
-        self.master2, self.master2, self.master2)
+         self.master2, self.master2, self.master2)
```
Please revert! The indentation was right before, plus it is unrelated.

lukpueh commented on tests/test_verifylib.py:
```diff
@@ -1196,7 +1171,7 @@ def test_verify_link_signature_thresholds__M_M_S__M_S_M__M_S_S(self):
     # see table in docstring of testcase
     signature = in_toto.gpg.functions.gpg_sign_object(
-        b"data", self.master, self.gnupg_home)
+         b"data", self.master, self.gnupg_home)
```
Please revert! The indentation was right before, plus it is unrelated.

lukpueh commented on tests/test_verifylib.py:
```diff
@@ -1284,12 +1256,10 @@ def setUpClass(self):
-    # Copy demo files to temp dir
-    for file in os.listdir(demo_files):
+    # Copy demo files to temp dir	
+    for file in os.listdir(demo_files):	
```
Remove trailing whitespace

lukpueh commented on tests/test_verifylib.py:
```diff
@@ -644,6 +638,8 @@ class TestInTotoVerify(unittest.TestCase):
 
   """
+  extra_settings = "demo"
+
```
Please remove. `extra_setting` is no longer used.

lukpueh reviewed (changes requested):
Thanks! Please address comments, and adopt your new class in other test scripts too, if applicable.

Also, please always proof-read your diffs carefully before requesting a review, and pay extra attention to things you have been asked before (e.g. [avoiding trailing whitespace](https://github.com/in-toto/in-toto/pull/216#pullrequestreview-222209928), [double indentation for line continuation](https://github.com/in-toto/in-toto/pull/216#discussion_r271751836), [advice for docstrings](https://github.com/in-toto/in-toto/pull/216#discussion_r271745008), [going through all test_*.py files](https://github.com/in-toto/in-toto/pull/216#issuecomment-480202219), ...).

michizhou committed:
Integrated new test scripts for modified refactoring

Incorporated additional test scripts into refactoring for setup and teardown of
temporary directory environments and modified refactoring implementation

lukpueh commented:
Thanks for you contribution, @michizhou. I'm closing this in favor of #345.
